Add vtk.write capabilities

kbarnhart opened the pull request:
As part of the recent hackathon I have added vtk.write to pyclaw.fileio. This work built on the substantial effort of @xinshengqin from [this branch](https://github.com/clawpack/visclaw/tree/dev_other_vis_packages) of visclaw. 

I took the code from the visclaw branch and gave it the same interface as the other write functions in fileio. I expanded the docstring, discussed the files that are written out, and included a few key paraview notes. I also modified the vtk.write function so that output can be placed in a generic directory (as opposed to the same directory as the function). 

I made a test that does a line-by-line comparison with some sample output. Since there is no vtk.read function, this seemed to me to be the best approach to testing. Other ideas for testing are welcome. I verified that my sample test output can be read in by paraview as well. 

Tests pass locally. 
I did a minor delint based on the output of pylint.  

I welcome feedback to improve/modify this PR to fit clawpack standards.

mandli commented:
A couple of initial comments:
1. Do we need all of the test files or can we use just one?
1. The only thing testing wise I would suggest is perhaps to use the python vtk module?

That last point did make me wonder if some of the code could be replaced by use of some of the functionality in that module.

ketch commented on src/petclaw/tests/test_io.py:
```diff
@@ -21,4 +21,7 @@ def test_data_dir(self):
\ No newline at end of file
+        return
+
+    def test_io_to_vtk(self):
```
Actually VTK output is already in petclaw (just grep for vtk in the petclaw directory).

kbarnhart replied:
Found it! Thanks for pointing this out.

kbarnhart replied:
@ketch I have a clarification question on this point. 

The vtk test that I've written relies on reading in a binary file and then verifying that the output written is the same as a sample vtk file. I wrote the test this way because there was a simple binary file in the test data folder already, and because this allows me to test the vtk output without a vtk writer (would be nice, but not a priority at the moment). 

When I remove the redefinition I get the following message:
```
ERROR: test_io.PetClawIOTest.test_io_to_vtk
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/barnhark/opt/anaconda3/envs/clawpack/lib/python3.8/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/Users/barnhark/projects/clawpack/pyclaw/src/pyclaw/tests/test_io.py", line 76, in test_io_to_vtk
    ref_sol.read(0,path=regression_dir,file_format="binary")
  File "/Users/barnhark/projects/clawpack/pyclaw/src/pyclaw/solution.py", line 343, in read
    read_func = self.get_read_func(file_format)
  File "/Users/barnhark/projects/clawpack/pyclaw/src/petclaw/solution.py", line 17, in get_read_func
    raise ValueError("File format %s not supported." % file_format)
ValueError: File format binary not supported.
```
Because my test relies on reading binary, and this error indicated to me that petclaw can't read binary, I redefined the function, just as `test_io_from_binary` was redefined. As such, I think that it makes sense to retain this redefinition. 

If you have other recommendations for how to handle this, I'd be happy to hear them.

ketch commented:
Thanks!  The existing VTK output (in PetClaw) is only for non-AMR grids, so this is a major extension.  On the question of outsourcing some of this to VTK (for perhaps easier maintainability), I noticed that the VTK docs at https://vtk.org/Wiki/VTK/Writing_VTK_files_using_python suggest looking at this package:

https://bitbucket.org/pauloh/pyevtk/src/default/

It looks like that makes it trivial to write a VTK file for a single numpy array, but I'm not sure how useful it would be for dealing with AMR grids.

kbarnhart commented:
Thanks for the comments @mandli and @ketch.  A couple of follow up comments and questions:

1) @mandli You are correct that I don't need to use an example with 17 grids, 2 is probably plenty. I used this particular example because it was already being used in the i/o examples, so the files to read in already existed. I can think of two options: 
   a) use a different example to read in, so that only 2 grids get written out
   b) only compare 2 of the 17 grids that write.vtk will write out. 

    Do you have any preference? If you think (a) is the better choice, I'd happily take recommendations on what clawpack example to use. (b) is super easy to implement. 

2) Regarding use of existing python packages, I did some (but not as extensive as I'd like) research into whether an existing python package would work for this application. I had hoped to find an existing package that could support both reading and writing. One option is the full VTK package from Kitware ([link to python wrapper in their source code](https://gitlab.kitware.com/vtk/vtk/-/tree/master/Wrapping/Python)). This would provides support from all VTK classes ([blog post example on improving numpy integration that mentions AMR](https://blog.kitware.com/improved-vtk-numpy-integration-part-5/)), but it would require adding that dependency  ([python wheels are here](https://vtk.org/download/), [conda-forge too](https://anaconda.org/conda-forge/vtk)). In contrast, the python package @ketch mentions looks like the best pure python thing available, but doesn't have support for the [vtkOverlappingAMR](https://vtk.org/doc/nightly/html/classvtkOverlappingAMR.html) class. 

     Because it didn't make sense to me to add the full VTK package dependency when @xinshengqin had written a writer for the single needed VTK class, I went with that option. 

    I'm not sure what the *best* approach is in this case, and am definitely open to input. If you all think that adding the vtk dependency makes the most sense, I think I would be able to modify this PR to do the write portion of i/o. I'm still working to understand the structure of the solution object, so I might need some additional input on the write portion.

kbarnhart commented:
The more I think about this, the more I think that the right approach is to try and use the core vtk python package. Its distributed through conda-forge, so managing the dependency should not be difficult. I'm going to take a stab at doing that. I'll report back as I progress and if I encounter issues.

mandli commented:
@kbarnhart Ah, I did not realize that this was one file/grid rather than a number of different time stamps.  In that case I think this is fine the way it is.

kbarnhart commented:
I've refactored this to use the vtk package directly from kitware. I have not made a vtk.read function. This would be nice, but I can't prioritize it at this point. 

It looks like this passes CI on py3.6, but not on py2.7. Looking at the travis logs it appears that that the errors are independent of this PR, but I'm not super familiar with the pyclaw CI. Let me know if anyone has ideas for how to resolve these failed builds and I can work on it. 

This introduces a new dependency (available through both conda-forge and pip). I've added this dependency to the pyclaw .travis.yml file. 

Its not entirely clear to me how/where to indicate this new dependency elsewhere in pyclaw and clawpack. I usually would edit the setup.py file (or requirements.txt, or environment.yml, or similar) to include new dependencies. However, it looks like the setup.py file doesn't include any of what I might call "optional" dependencies (e.g., hdf). 

If there are other places that I need to indicate the dependency, please let me know so I can make those changes.

rjleveque commented:
@kbarnhart Thanks for working on this!  @mandli or @ketch may have other suggestions, but one place we should add new dependencies is to https://github.com/clawpack/docker-files.  The most recent one is [Dockerfile_v5.7.0_geoclaw](https://github.com/clawpack/docker-files/blob/master/Dockerfile_v5.7.0_geoclaw).

We are officially [dropping support for Python 2.7 ](http://www.clawpack.org/python.html) as of v5.7.0, so it may be time to stop testing this on travis.

ketch commented:
@kbarnhart All I meant was to point out that the written comment in the code is potentially misleading.  Of course, "binary" is not a file format but just shorthand within pyclaw for referring to a certain non-ASCII output format.  You don't need to write a test for PetClaw VTK output; the PetClaw output formats are already tested.

kbarnhart commented:
thanks @rjleveque I've created a PR to add the dependency to the dockerfile. clawpack/docker-files#9

@ketch thanks for the clarification. 

Happy to continue to revise this based this based on your comments.

kbarnhart commented:
I wanted to check in and see if there was anything else I could do on this PR.

mandli commented:
@kbarnhart Does `vtkOverlappingAMR` exist for the version of VTK that you are using?

kbarnhart commented on src/pyclaw/fileio/vtk.py:
```diff
@@ -0,0 +1,317 @@
+
+    # Initialize the vtkOverlappingAMR object. Provide it the number of levels,
+    # number of blocks per level, and the global origin.
+    amr = vtk.vtkOverlappingAMR()
```
@mandli yup! I used the kitware distributed vtk python package, which was definitely the right choice. 

Making a comment here so its easy to see where the vtkOverlappingAMR object is defined.

mandli replied:
Is there a way to use that distribution then for the testing? Alternatively this will always fail and we need to remove the test unfortunately.

kbarnhart replied:
I now see that the CI was re-run and there is an issue that in py27 (which uses vtk v8.2) this class can't be found. Last time I looked at the CI it had failed at an earlier place so I hadn't seen this (thanks!) 

The py36 CI also uses vtk v8.2, so I'm hopeful that this will be fixable. Will try to reproduce this error and report back. 

Thanks!

kbarnhart commented:
Alright. Looks like this was a python 27 namespace issue which I have now fixed. The [py36 petclaw test fails](https://travis-ci.org/github/clawpack/pyclaw/jobs/704423211) and it appears to be independent to me (though I may be mistaken). 

```
======================================================================
ERROR: examples.psystem_2d.psystem_2d({'solver_type': 'classic', 'outdir': './_for_temp_pyclaw_test', 'cells_per_layer': 6, 'tfinal': 1.0})
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/miniconda/lib/python3.6/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/miniconda/lib/python3.6/site-packages/clawpack/pyclaw/util.py", line 174, in <lambda>
    test = lambda: test_app(application, verifier, test_kwargs)
  File "/home/travis/miniconda/lib/python3.6/site-packages/clawpack/pyclaw/util.py", line 240, in test_app
    check_values = verifier(claw)
  File "/home/travis/build/clawpack/pyclaw/examples/psystem_2d/test_2d_psystem.py", line 36, in verify
    reltol=1e-4)
  File "/home/travis/miniconda/lib/python3.6/site-packages/clawpack/pyclaw/util.py", line 275, in check_diff
    err_norm = d*np.linalg.norm(expected - test)
ValueError: operands could not be broadcast together with shapes (24,3) (0,) 
-------------------- >> begin captured stdout << ---------------------
{'use_petsc': True, 'kernel_language': 'Fortran', 'solver_type': 'classic', 'outdir': './_for_temp_pyclaw_test', 'cells_per_layer': 6, 'tfinal': 1.0}
gauge directory already exists, ignoring
directory already exists, ignoring
--------------------- >> end captured stdout << ----------------------
-------------------- >> begin captured logging << --------------------
root: DEBUG: No netcdf4 support found.
--------------------- >> end captured logging << ---------------------
----------------------------------------------------------------------
Ran 86 tests in 1664.082s
FAILED (errors=1)
```

mandli commented:
Trying to restart. Sometimes these things fail for unexpected reasons.

ketch commented:
This test seems to be failing intermittently on master too.  I'm not sure why, but I don't think it has to do with this PR so we don't need to let it stand in the way of merging this.

mandli commented:
The version that failed this time changed so I tend to agree with @ketch that this is not a problem with this PR.

kbarnhart commented:
@mandli @ketch thanks! I don't know if there is anything else I need to do for this right now, but if something comes up, let me know and I'll work on it.

mandli commented:
We will try just restarting the failed version.  If that still does not work let's merge this and try and debug travis.

rjleveque commented:
@mandli, @ketch, @kbarnhart: travis tests seem to be passing, is this ready to merge?

ketch commented:
Yes, let's merge it.  If @kbarnhart can also add something about this to the docs (in a PR to clawpack/doc) that would be great.

kbarnhart commented:
@ketch I've put making a doc PR on my todo list. Might be next week, but will do.

ketch commented:
@kbarnhart Thanks!
